File: 프로그래머스/2/154540. 무인도 여행/무인도 여행.py
def solution(maps):
    def bfs(i, j):
        total = int(maps[i][j])
        visited[i][j] = True
        need_visit = [(i,j)]
        
        while need_visit:
            tmp = need_visit.pop(0)
            r = tmp[0]
            c = tmp[1]
            for k in range(4):
                nr = r + di[k]
                nc = c + dj[k]
                if 0 <= nr < m and 0 <= nc < n and (not visited[nr][nc]) and maps[nr][nc] != 'X':
                    visited[nr][nc] = True
                    total += int(maps[nr][nc])
                    need_visit.append((nr, nc))
        
        return total
        
    answer = []
    di = [0, 0, 1, -1]
    dj = [1, -1, 0, 0]
    
    m = len(maps)
    n = len(maps[0])
    
    visited = [[False] * n for _ in range(m)]

    for i in range(m):
        for j in range(n):
            if maps[i][j] != 'X' and not visited[i][j]:
                total = bfs(i, j)
                answer.append(total)
                
    answer.sort()
    
    if not answer:
        answer = [-1]
    
    return answer

File: 프로그래머스/2/154540. 무인도 여행/무인도 여행.js
function solution(maps) {
    var answer = [];
    
    
    // 가로 세로 길이
    const m = maps.length
    const n = maps[0].length
    
    let newArray = [];
    let visited = [];
    

    // 숫자, 배열로 변환
    maps.map((row) => {
        let newRow = [];    // 새로운 배열 값
        let tmpRow = [];    // visited 배열 값

        const listRow = [...row]
        
        listRow.map((col) => {
            if (col !== 'X') {
                newRow.push(Number(col))
                tmpRow.push(false)
            }
            else {
                newRow.push(col)
                tmpRow.push(false)
            }
        })
        
        newArray.push(newRow)
        visited.push(tmpRow)
    })
    
    
    // 델타 탐색, bfs 함수 정의
    const di = [0, 0, 1, -1]
    const dj = [1, -1, 0, 0]
    
    function bfs(i, j) {
        let total = newArray[i][j]
        visited[i][j] = true
        let needVisit = [[i, j]];
        
        while (needVisit.length > 0) {
            const tmp = needVisit.shift();
            const r = tmp[0]
            const c = tmp[1]
            for (let k = 0; k < 4; k++) {
                const nr = r + di[k]
                const nc = c + dj[k]
                if ((0 <= nr && nr < m) && (0 <= nc && nc < n) && !visited[nr][nc] && newArray[nr][nc] != 'X') {
                    visited[nr][nc] = true
                    total += newArray[nr][nc]
                    needVisit.push([nr, nc])
                    
                }
            }
            
        }
        
        return total
    }
    
    
    // 순회하면서 bfs 수행하기
    for (let a = 0; a < m; a++) {
        for (let b = 0; b < n; b++) {
            if (newArray[a][b] !== 'X' && !visited[a][b]) {
                const total = bfs(a, b)
                answer.push(total)
            }
        }
    }
    
    answer.sort((a, b) => a - b)
    
    if (answer.length === 0) {
        answer = [-1]
    }

    
    return answer;
}